File: frontend/src/api/audit.js
import { apiRequest } from './client';

export const listAuditEntries = (limit = 50, offset = 0) => apiRequest(`/api/audit?limit=${limit}&offset=${offset}`);

const byOrdinal = (left, right) => left.sequenceOrdinal - right.sequenceOrdinal;

export const mapProfessorContentAudit = (audit = {}) => {
	const responses = (audit.responses || []).map((response) => ({
		id: response.id,
		sequenceEntryId: response.sequence_entry_id,
		responseVersion: response.response_version,
		kind: response.command_kind,
		answer: response.normalized_answer,
		answeredAt: response.answered_at,
		requestDigest: response.request_digest,
		idempotencyKeyDigest: response.idempotency_key_digest,
	}));
	const effects = (audit.effects || []).map((effect) => ({
		id: effect.id,
		sequenceEntryId: effect.sequence_entry_id,
		effectIndex: effect.effect_index,
		payload: effect.effect_payload,
		beforeProjectionDigest: effect.before_projection_digest,
		afterProjectionDigest: effect.after_projection_digest,
		appliedAt: effect.applied_at,
		turnId: effect.turn_id,
		turnWeekNumber: effect.turn_week_number,
	}));
	const deliveries = (audit.deliveries || []).map((delivery) => ({
		id: delivery.id,
		sequenceEntryId: delivery.sequence_entry_id,
		sequenceOrdinal: delivery.sequence_ordinal,
		checkpoint: delivery.checkpoint,
		visibility: delivery.visibility,
		hiddenFromStudents: delivery.hidden_from_students,
		definitionDigest: delivery.definition_digest,
		definition: delivery.definition_snapshot,
		status: delivery.status,
		deliveredAt: delivery.delivered_at,
		completedAt: delivery.completed_at,
		turnId: delivery.turn_id,
		turnWeekNumber: delivery.turn_week_number,
		responses: responses.filter((response) => response.sequenceEntryId === delivery.sequence_entry_id),
		effects: effects.filter((effect) => effect.sequenceEntryId === delivery.sequence_entry_id),
	})).sort(byOrdinal);

	return {
		deliveries,
		responses,
		effects,
		digestStatus: audit.digest_status || 'unknown',
		divergences: (audit.divergences || []).map((divergence) => ({
			category: divergence.category,
			record: divergence.record,
			expected: divergence.expected,
			actual: divergence.actual,
		})),
	};
};

export const mapRunAudit = (result) => ({
	...result,
	contentAudit: mapProfessorContentAudit(result.content_audit),
});

File: backend/app/api/audit.py
from datetime import datetime

from fastapi import APIRouter, Query
from pydantic import BaseModel, ConfigDict

from app.api.auth import ProfessorUser
from app.api.classes import DatabaseSession
from app.audit.service import list_actor_audit

router = APIRouter(prefix="/audit", tags=["audit"])


class AuditResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    action: str
    target_type: str
    target_id: str
    details: dict[str, object]
    created_at: datetime


class ContentDeliveryAuditResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    sequence_entry_id: str
    sequence_ordinal: int
    checkpoint: str
    visibility: str
    hidden_from_students: bool
    definition_digest: str
    definition_snapshot: dict[str, object]
    status: str
    delivered_at: datetime
    completed_at: datetime | None
    turn_id: str | None
    turn_week_number: int | None


class ContentResponseAuditResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    sequence_entry_id: str
    response_version: int
    command_kind: str
    normalized_answer: dict[str, object]
    answered_at: datetime
    request_digest: str
    idempotency_key_digest: str


class AppliedEffectAuditResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    sequence_entry_id: str
    effect_index: int
    effect_payload: dict[str, object]
    before_projection_digest: str
    after_projection_digest: str
    applied_at: datetime
    turn_id: str | None
    turn_week_number: int | None


class ReplayDivergenceResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    category: str
    record: str
    expected: object
    actual: object


class ProfessorContentAuditResponse(BaseModel):
    deliveries: list[ContentDeliveryAuditResponse]
    responses: list[ContentResponseAuditResponse]
    effects: list[AppliedEffectAuditResponse]
    digest_status: str
    divergences: list[ReplayDivergenceResponse]


@router.get("", response_model=list[AuditResponse])
def get_audit_history(
    session: DatabaseSession,
    professor: ProfessorUser,
    limit: int = Query(default=50, ge=1, le=200),
    offset: int = Query(default=0, ge=0),
) -> object:
    return list_actor_audit(session, actor_id=professor.id, limit=limit, offset=offset)
